nhs/data/geography.py

```diff
@@ -0,0 +1,295 @@
+        area_code (str): The area code to which shuffled coordinates should be assigned.
+        shuffled_coords (List[Tuple[float, float]]): List of shuffled coordinate pairs.
+
+    Returns:
+        pl.DataFrame: Dataframe with shuffled coordinates assigned for the specified area_code.
+    """  # noqa: E501
+    return df.with_columns(
+        [
+            pl.when(pl.col("area_code") == area_code)
+            .then(pl.Series([coord[0] for coord in shuffled_coords]))
+            .otherwise(pl.col("x"))
+            .alias("x"),
+            pl.when(pl.col("area_code") == area_code)
+            .then(pl.Series([coord[1] for coord in shuffled_coords]))
+            .otherwise(pl.col("y"))
+            .alias("y"),
+        ]
+    )
+
+
+def random_distribution(
+    df: pl.DataFrame, num_iterations: int = 10, seed: int = 42
+) -> pl.DataFrame:
+    """
+    Randomly shuffle lat-long pairs within their respective areas, maintaining uniqueness.
+
+    This function shuffles coordinates within each area_code, ensuring that the number of unique
+    coordinates remains the same before and after shuffling. It uses a seed for reproducibility.
+
+    This was used in testing for Schelling's model of segregation, which worked with very limited
+    success (something like 30minutes to move 5 people).
+
+    Remains here in case of future expansion into distributions.
+
+    Args:
+        df (pl.DataFrame): Input dataframe containing 'x', 'y', and 'area_code' columns.
+        num_iterations (int, optional): Number of shuffling iterations. Defaults to 10.
+        seed (int, optional): Seed for the random number generator. Defaults to 42.
+
+    Returns:
+        pl.DataFrame: Dataframe with shuffled coordinates, maintaining original uniqueness within each area_code.
+    """  # noqa: E501
+    model_df = df.clone()
+    model_df = model_df.with_columns(
+        [pl.col("y").alias("original_lat"), pl.col("x").alias("original_lon")]
+    )
+
+    area_codes = model_df["area_code"].unique().to_list()
+
+    for _ in range(num_iterations):
+        for area_code in area_codes:
+            area_df = model_df.filter(pl.col("area_code") == area_code)
+            unique_coords = area_df[["x", "y"]].unique().to_numpy().tolist()
+            shuffled_coords = shuffle_coordinates(unique_coords, seed)
+
+            # Create a mapping from original to shuffled coordinates
+            coord_map = dict(zip(unique_coords, shuffled_coords))
+
+            # Apply the mapping to maintain uniqueness
+            model_df = model_df.with_columns(
+                [
+                    pl.when(pl.col("area_code") == area_code)
+                    .then(
+                        pl.struct(["x", "y"]).map_elements(
+                            lambda c: coord_map.get((c["x"], c["y"]), (c["x"], c["y"]))
+                        )
+                    )
+                    .otherwise(pl.struct(["x", "y"]))
+                ]
+            ).unnest("literal")
+
```

Comment:
Can move to its own function to avoid nested for loops. `model_df_with_columns` is also very heavily nested, maybe make intermediate variables to make this easier to read and less nested since it's quite hard to read